### PR TITLE
Add Parser for Generic Renewable Models

### PR DESCRIPTION
**Pull request by rodrigomha**
To do:
- [x] Add Test. Requires adding a new .dyr file to the `PowerSystemsTestData` repo, and maybe modifying the CaseBuilder repo.
- [x] Not sure if we need a keyword constructor for the new constructor of the `DynamicInverter` that splits the `OuterControl` into the Active and Reactive part.

### CODE REVIEW

File: src/models/dynamic_inverter.jl [hunk 1]
```diff
@@ -171,6 +171,63 @@ function DynamicInverter(;
...
+    states = _calc_states(
+        converter,
+        outer_control,
+        inner_control,
+        dc_source,
+        freq_estimator,
+        filter,
+    )
+
+    return DynamicInverter{C, O, IC, DC, P, F}(
```
Comment (jd-lara) [inline src/models/dynamic_inverter.jl:214]: You don't need the types in this constructor call. 
```suggestion
    return DynamicInverter(
```

File: src/models/dynamic_inverter.jl [hunk 2]
```diff
@@ -171,6 +171,63 @@ function DynamicInverter(;
...
+        filter,
+    )
+    states = _calc_states(
+        converter,
+        outer_control,
+        inner_control,
+        dc_source,
+        freq_estimator,
+        filter,
+    )
```
Comment (jd-lara) [inline src/models/dynamic_inverter.jl:212]: These fields don't need to be defined here. Calling the correct constructor will populate the fields automatically

Comment (rodrigomha) [reply src/models/dynamic_inverter.jl]: Properly fixed calling directly the function

**Comment by jd-lara** [comment]
@rodrigomha you can use this branch https://github.com/NREL-SIIP/PowerSystemCaseBuilder.jl/pull/24 to add the tests locally. The system is built with the function `sys = build_system(PSYTestSystems, "psse_renewable_parsing_1")`

File: src/parsers/psse_dynamic_data.jl [hunk 3]
```diff
@@ -44,15 +44,46 @@ function _populate_args(param_map, val)
...
+        #If the parameter is a Float64, then use the value directly as the argument.
+        #Typically uses for the resistance that is not available in .dyr files.
+        if isa(_v, Float64)
+            struct_args[ix] = _v
+            #If the parameter is an Int, then use the integer as the key of the value in the dictionary.
+        elseif isa(_v, Int)
+            struct_args[ix] = val[_v]
+            #If the parameter is a tuple (as a string), then construct the tuple directly.
+        elseif occursin(r"^\(\d+\s*,\s*\d+\)", _v) #is a tuple string
+            _t = strip(_v, ['(', ')'])
```
Comment (daniel-thom) [inline src/parsers/psse_dynamic_data.jl:71]: You could make a slight improvement by capturing the matches like this:
```
julia> m = match(r"^\((\d+)\s*,\s*(\d+)\)$", "(1, 2)")
RegexMatch("(1, 2)", 1="1", 2="2")

julia> if m !== nothing
           @show parse.(Int, m.captures)
       end
parse.(Int, m.captures) = [1, 2]
2-element Vector{Int64}:
 1
 2
```

File: src/parsers/psse_dynamic_data.jl [hunk 4]
```diff
@@ -163,15 +247,121 @@ function _parse_dyr_components(data::Dict)
...
+
+"""
+Convert specific parameters to types that are not Float64 for
+specific inverter components
+
+"""
+function _convert_argument_types!(str::AbstractString, struct_args::Vector)
+    if str == "ActiveRenewableControllerAB"
+        struct_args[1:5] .= Int.(struct_args[1:5])
+        struct_args[4] = string(struct_args[4])
```
Comment (daniel-thom) [inline src/parsers/psse_dynamic_data.jl:354]: Was this already a string when entering the function?

File: src/models/dynamic_inverter.jl [hunk 5]
```diff
@@ -171,6 +171,38 @@ function DynamicInverter(;
...
+function DynamicInverter(
+    name::String,
+    ω_ref::Float64,
+    converter::C,
+    active_power_control::AC,
+    reactive_power_control::RC,
+    inner_control::IC,
+    dc_source::DC,
+    freq_estimator::P,
+    filter::F,
```
Comment (daniel-thom) [inline src/models/dynamic_inverter.jl:183]: Recommend against overriding standard library functions.

File: src/parsers/psse_dynamic_data.jl [hunk 6]
```diff
@@ -233,6 +423,12 @@ function add_dyn_injectors!(sys::System, bus_dict_gen::Dict)
             #Add Dynamic Generator
             dyn_gen = DynamicGenerator(get_name(g), 1.0, temp_dict[_id]...)
             add_component!(sys, dyn_gen, g)
+        elseif length(temp_dict[_id]) == 7 #Inverter has 7 components (6 if Outer is put together)
+            #To do: Check X_Source and R_Source data if we want
```
Comment (jd-lara) [inline src/parsers/psse_dynamic_data.jl:427]: I think we need to address this from the data to properly build the case.

File: src/parsers/psse_dynamic_data.jl [hunk 7]
```diff
@@ -44,15 +50,50 @@ function _populate_args(param_map, val)
         elseif isa(_v, Int)
             struct_args[ix] = val[_v]
             #If the parameter is a tuple (as a string), then construct the tuple directly.
-        else
+        elseif occursin(r"^\(\d+\s*,\s*\d+\)", _v) #is a tuple string
```
Comment (daniel-thom) [inline src/parsers/psse_dynamic_data.jl:53]: Do you want to apply the improved regex here also?

**Comment by jd-lara** [comment]
Tests are expected to fail until a new version of PSB with the test systems is released

File: src/parsers/psse_dynamic_data.jl [hunk 8]
```diff
@@ -163,15 +253,120 @@ function _parse_dyr_components(data::Dict)
                     temp[component_table[gen_field]] = component_constructor(struct_args...)
                 end
             else
-                @warn "$(componentID[1]) at bus $bus_num, id $(componentID[2]), not supported in PowerSystems.jl. Skipping data."
+                #@warn "$(componentID[1]) at bus $bus_num, id $(componentID[2]), not supported in PowerSystems.jl. Skipping data."
```
Comment (daniel-thom) [inline src/parsers/psse_dynamic_data.jl:256]: Is this inadvertent?

**Review by daniel-thom — APPROVED** [review]

**Review by jd-lara — APPROVED** [review]
We can merge once PSB is ready and the CI runs

**Pull request merged**